json_object.c

```diff
@@ -1202,49 +1303,75 @@ const char *json_object_get_string(struct json_object *jso)
 	default: return json_object_to_json_string(jso);
 	}
 }
-
 int json_object_get_string_len(const struct json_object *jso)
 {
+	ssize_t len;
 	if (!jso)
 		return 0;
 	switch (jso->o_type)
 	{
-	case json_type_string: return jso->o.c_string.len;
+	case json_type_string:
+		len = JC_STRING_C(jso)->len;
+		return (len < 0) ? -(ssize_t)len : len;
 	default: return 0;
 	}
 }
 
-int json_object_set_string(json_object *jso, const char *s)
-{
-	return json_object_set_string_len(jso, s, (int)(strlen(s)));
-}
-
-int json_object_set_string_len(json_object *jso, const char *s, int len)
+static int _json_object_set_string_len(json_object *jso, const char *s, size_t len)
 {
 	char *dstbuf;
+	ssize_t curlen;
+	ssize_t newlen;
 	if (jso == NULL || jso->o_type != json_type_string)
 		return 0;
-	if (len < LEN_DIRECT_STRING_DATA)
-	{
-		dstbuf = jso->o.c_string.str.data;
-		if (jso->o.c_string.len >= LEN_DIRECT_STRING_DATA)
-			free(jso->o.c_string.str.ptr);
-	}
-	else
+
+	if (len >= SSIZE_T_MAX - 1)
+		// jso->len is a signed ssize_t, so it can't hold the
+		// full size_t range.
+		return 0;
+
+	dstbuf = get_string_component_mutable(jso);
+	curlen = JC_STRING(jso)->len;
+	if (curlen < 0)
+		curlen = -curlen;
+	newlen = len;
+
+	if ((ssize_t)len > curlen)
```

Comment:
Due to the previous memory allocation, The logic here becomes more complicated. But considering that this function is used less, I think it's acceptable.

Reply:
> 
> 
> Sorry for the delay in the review.

No worries!  Thanks for taking a look.